pinot-minion-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/CustomerBasedRetentionTaskGenerator.java

```diff
@@ -0,0 +1,162 @@
+
+  private ClusterInfoAccessor _clusterInfoAccessor;
+
+  @Override
+  public void init(ClusterInfoAccessor clusterInfoAccessor) {
+    _clusterInfoAccessor = clusterInfoAccessor;
+  }
+
+  @Override
+  public String getTaskType() {
+    return TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+
+    // Get the segments that are being converted so that we don't submit them again
+    Set<Segment> runningSegments =
+        TaskGeneratorUtils.getRunningSegments(MinionConstants.ConvertToRawIndexTask.TASK_TYPE, _clusterInfoAccessor);
+
+    for (TableConfig tableConfig : tableConfigs) {
+      String offlineTableName = tableConfig.getTableName();
+
+      if (tableConfig.getTableType() != TableType.OFFLINE) {
+        LOGGER.warn("Skip generating task: {} for non-OFFLINE table: {}", TASK_TYPE, offlineTableName);
+        continue;
+      }
+
+      LOGGER.info("Start generating task configs for table: {} for task: {}", offlineTableName, TASK_TYPE);
+
+      // Only schedule 1 task of this type, per table
+      Map<String, TaskState> incompleteTasks =
+          TaskGeneratorUtils.getIncompleteTasks(TASK_TYPE, offlineTableName, _clusterInfoAccessor);
+      if (!incompleteTasks.isEmpty()) {
+        LOGGER
+            .warn("Found incomplete tasks: {} for same table: {}. Skipping task generation.", incompleteTasks.keySet(),
+                offlineTableName);
+        continue;
+      }
+
+      TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
+      Preconditions.checkState(tableTaskConfig != null);
+      Map<String, String> taskConfigs = tableTaskConfig.getConfigsForTaskType(TASK_TYPE);
+      Preconditions.checkState(taskConfigs != null, "Task config shouldn't be null for table: {}", offlineTableName);
+
+      // Get the bucket size
+      String bucketTimePeriod =
+          taskConfigs.getOrDefault(BUCKET_TIME_PERIOD_KEY, DEFAULT_BUCKET_PERIOD);
+      long bucketMs = TimeUtils.convertPeriodToMillis(bucketTimePeriod);
+
+      // Get watermark from OfflineSegmentsMetadata ZNode. WindowStart = watermark. WindowEnd = windowStart + bucket.
+      long windowStartMs = getWatermarkMs(offlineTableName, bucketMs);
+      long windowEndMs = windowStartMs + bucketMs;
+
+      // Get max number of tasks for this table
+      int tableMaxNumTasks = getTableMaxNumTasks(taskConfigs);
+
+      // Get customer retention config
+      Map<String ,String> customerRetentionConfigMap = getCustomerRetentionConfig();
+      String customerRetentionConfigMapString = customerRetentionConfigMap.keySet().stream()
+          .map(key -> key + "=" + customerRetentionConfigMap.get(key))
+          .collect(Collectors.joining(", ", "{", "}"));
+
+      // Generate tasks
+      int tableNumTasks = 0;
+      for (OfflineSegmentZKMetadata offlineSegmentZKMetadata : _clusterInfoAccessor.getOfflineSegmentsMetadata(offlineTableName)) {
```

Comment:
This api returns ordered results in terms of timestamp? In this case, we need ordered processing I believe.

Reply:
Could you explain how does the order mater over here ?

Reply:
resolved